igvm/hypervisor.py

```diff
@@ -800,34 +885,143 @@ def check_netcat(self, port):
             )
 
     def kill_netcat(self, port):
-        self.run('pkill -f "^/bin/nc.openbsd -l -p {}"'.format(port))
+        self.run(
+            'pkill -f "^/bin/nc.openbsd -l -p {}"'.format(port),
+            warn_only=True,  # It's fine if the process already dead
+        )
 
-    @contextmanager
-    def netcat_to_device(self, device):
+    def _netcat_port(self, device: str) -> int:
+        """
+        Get the minor ID for the device, calculates the netcat listen port for
+        it, checks if netcat process already except, and returns the port
+        """
         dev_minor = self.run('stat -L -c "%T" {}'.format(device), silent=True)
         dev_minor = int(dev_minor, 16)
         port = 7000 + dev_minor
-
         self.check_netcat(port)
+        return port
+
+    @contextmanager
+    def netcat_to_device(self, device: str) -> Iterator[Tuple[str, int]]:
+        """
+        Spawns the backgroung netcat process on the uniq port and pipes the
+        payload to dd process to restore the file system on a target
+        hypervisor. Kills the netcat process if exceprion is cought.
+
+        :param str device: The path to the volume device
+        :rtype Iterator[Tuple[str, int]]: (fqdn, port) pair
+        """
+        port = self._netcat_port(device)
 
         # Using DD lowers load on device with big enough Block Size
         self.run(
             'nohup /bin/nc.openbsd -l -p {0} | dd of={1} obs=1048576 &'
-            .format(port, device)
+            .format(port, device),
+            pty=False,  # Has to be here for background processes
         )
         try:
             yield self.fqdn, port
         except BaseException:
             self.kill_netcat(port)
             raise
 
-    def device_to_netcat(self, device, size, listener):
+    def device_to_netcat(
+        self, device: str, size: int, listener: Tuple[str, int]
+    ):
+        """
+        Dumps the device via dd and netcat to a remote listener
+
+        :param str device: The path to the volume device
+        :param int size: The disk size for pv progress and ETA
+        :listener Tuple[str, int] listener: (fqdn, port) pair for nc connection
+        """
         # Using DD lowers load on device with big enough Block Size
         self.run(
             'dd if={0} ibs=1048576 | pv -f -s {1} '
             '| /bin/nc.openbsd -q 1 {2} {3}'
-            .format(device, size, *listener)
+            .format(device, size, listener[0], listener[1])
+        )
+
+    def _xfsrestore_log_name(self, vm: VM) -> str:
+        device_name = self.get_volume_by_vm(vm).name()
+        return '/tmp/xfsrestore-{}.log'.format(device_name)
+
+    def check_xfsrestore_log(self, vm: VM):
+        """
+        Search for WARNING in the xfsrestore log file.
+        Raises exception if found
+        """
+        self.run('cat {}'.format(self._xfsrestore_log_name(vm)))
```

Comment:
Why first we cat the file and then grep it?

Reply:
To have the log output in the migration log. Only for this.

Reply:
There is no migration log. If you mean dumping the output to  console, that's fine.